### DISCUSSION TITLE
Scaling GoCD for 1000+ pipelines

### DISCUSSION THREAD

**Discussion by phzietsman** [gocd/gocd#10073] [Q&A]
Hey there,

We've been using GoCD for about  9 months now and its been working like a charm, but we are seeing the service degrade over time. When we started the server was snappy and API calls got services in acceptable times, now we can easily wait a few seconds for a call.

We have roughly 1100 pipelines of which ~900 is created using the API and the rest sit in config repos and 110 unique materials that are being monitored. For our server the JVM heap max is set to 2G running in AWS ECS with EFS network storage for persistence.

Today it the default dashboard that shows all the pipelines takes about 3-4 seconds to load and anything that touches the config.xml takes even longer. 

In the steering [notes](https://github.com/gocd/steering/blob/master/notes/2020_07_08.md#discussions) it is mentioned that there are clients running GoCD with 4000 pipelines on a single instance. Can anybody share a setup (at a high level is fine) that can help us get to these numbers without the server getting sluggish.

thanks!

**Comment by chadlwilson** [comment]
Which version are you running? And what have you tried to improve performance?

Have you tried giving it more memory? 2G is not much for that many pipelines.

Also, how big is your EFS volume? IIRC you get different disk speeds for different volume sizes. I've only ever run it on EBS and latency is different there.

Usually people want to migrate to Postgres away from h2db at a certain scale and you may have hit that. I assume you are just using the default h2?

**Reply by phzietsman** [reply]
Hey Chad, thanks for the reply.

We are on 21.3. 

At one point we had a lot of unique materials for config repos, which we consolidated and that made a noticeable difference, but other than that nothing. 

We are using Postgres, we were worried about it with regards to performance since we knew we would be using GoCD a lot.

Regarding EFS, it has "unlimited" storage ( according to GoCD we have 7EB left ;) ) and we are on "General Purpose" Performance mode. EFS metrics indicate that we are not even utilizing 1% of the available IO. None the less we are going to test EBS.

We've also have 2 EFS volumes, one for artifacts and one just for config. We went with EFS due to its persistence, but it might be a better idea to use EBS as primary and EFS as a backup destination or just push backups to S3 directly.

Next steps for us:
1. Increase JVM memory to 8G
2. Move config to EBS
3. Move artifacts to EBS

From a performance point of view, is it better to use config repos or using the API to define pipeline?

**Reply by chadlwilson** [reply]
I would definitely try throwing memory and/or CPU cores at it first (on its own) and see how it goes. What do your metrics currently show about the container's CPU and memory usage? That seems likely to be a quicker experiment than changing storage. Other than that, you'd be looking at a way of profiling to determine what the problem is. While some of the information is a bit out of date and intended for non-cloud setups, the stuff on JVM heap profliing logs [here](https://docs.gocd.org/current/installation/performance_tuning.html) is probably relevant, although admittedly requiring some JVM expertise to interpret.

On EFS what I am referring to is that the size you actually store in EFS (and the cost) also affects [the throughput you get from it](https://docs.aws.amazon.com/efs/latest/ug/performance.html#throughput-modes), if not in "provisioned" mode. It's designed to give you more throughput/IOPS as your usage grows. If you only have say 10GiB on a volume with the config.xml you're only getting 100MiB/s throughput - while bursting. I'm not sure how that could affect things in your case, and it's probably not that big a deal, but seems worth being aware of with small EFS volumes? Nevertheless, in my experience, normally it is raw latency rather than throughput that is likely to be most relevant for sluggishness - I have only really seen that applying to those using h2. I don't personally have good intuitions for running it with Postgres since I haven't really done that before. Maybe some other folks can pitch in.

Do your postgres metrics indicate it also is under any particular pressure/contention/latency?

> From a performance point of view, is it better to use config repos or using the API to define pipeline?

I'm not sure, but personally I don't think that seems likely to make a much difference to performance, unless by "config repos" you are talking about adding another 100 repos that need to be polled. These are also "materials" in a sense.

Speaking of which, are your materials polling? Is switching to webhook notifications from your repo an option to experiment with if other things don't help out? Usually I'd only suggest that if you have seen very slow times for pipelines to trigger rather than "generalised UI slowness" as you seem to describe.

I also imagine if you dig back through GoCD Github issues or [the Google Group](https://groups.google.com/g/go-cd) you will likely find quite a few pieces of advice from the core team that are likely still valid. Don't have them on-hand, but worth a dig.
